### PR TITLE
[JVM] Improve debug step behavior around lambdas.

### PR DESCRIPTION
**Pull request by kandersen**
[JVM] Improve debug step behavior around lambdas. …

- Add tests to clarify problematic behavior
- Avoid line numbers on return instructions of lambdas without explicit returns

### CODE REVIEW

File: compiler/backend/src/org/jetbrains/kotlin/codegen/ExpressionCodegen.java [hunk 1]
```diff
@@ -1786,6 +1785,14 @@ private static boolean isLambdaVoidBody(@NotNull KtElement bodyExpression, @NotN
         return false;
     }
 
+    private static boolean isLambdaBody(@NotNull KtElement bodyExpression) {
```
Comment (madsager) [inline compiler/backend/src/org/jetbrains/kotlin/codegen/ExpressionCodegen.java:1788]: Maybe rewrite isLambdaVoidBody about to:

```
return isLambdaBody(bodyExpression) && Type.VOID_TYPE.equals(returnType);
```

File: compiler/testData/debug/stepping/voidLambdaStepInline.kt [hunk 2]
```diff
@@ -0,0 +1,20 @@
+
+
+
+
+// FILE: test.kt
+fun box(): String {
+    run { println("OK") }
```
Comment (madsager) [inline compiler/testData/debug/stepping/voidLambdaStepInline.kt:7]: I believe that we normally avoid having println calls in tests to not generate stdout output. Does this should the same if you call some other function?

File: compiler/testData/debug/stepping/lambdaStepInline.kt [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+// FILE: test.kt
+inline fun foo(stringMaker: () -> String): String {
+    return stringMaker()
+}
+
+fun box(): String {
+    foo { "OK "}
+    foo {
+        "OK"
+        // This line is hit by the old backend after line 13
```
Comment (pyos) [inline compiler/testData/debug/stepping/lambdaStepInline.kt:14]: Not anymore though :)

File: compiler/backend/src/org/jetbrains/kotlin/codegen/ExpressionCodegen.java [hunk 4]
```diff
@@ -1753,10 +1753,9 @@ public void returnExpression(@NotNull KtExpression expr) {
         if (!endsWithReturn(expr)) {
             if (isLambdaVoidBody(expr, typeForExpression)) {
                 markLineNumber((KtFunctionLiteral) expr.getParent(), true);
-            }
-            else {
+            } else if (!isLambdaBody(expr)){
                 markLineNumber(expr, true);
-            }
+            }g
```
[the remark this answers is not in the document]
Comment (kandersen) [reply compiler/backend/src/org/jetbrains/kotlin/codegen/ExpressionCodegen.java:1758]: Indeed!

File: compiler/backend/src/org/jetbrains/kotlin/codegen/ExpressionCodegen.java (hunk 1 above)
Comment (kandersen) [reply compiler/backend/src/org/jetbrains/kotlin/codegen/ExpressionCodegen.java:1788]: Good idea!

File: compiler/testData/debug/stepping/voidLambdaStepInline.kt (hunk 2 above)
Comment (kandersen) [reply compiler/testData/debug/stepping/voidLambdaStepInline.kt:7]: Not quite the same, but it still illustrates the issue that Denis pointed out in the chat last night - that there really should be two steps at 7.

File: compiler/testData/debug/stepping/lambdaStepInline.kt [hunk 5]
```diff
@@ -0,0 +1,26 @@
+
```
[the remark this answers is not in the document]
Comment (kandersen) [reply compiler/testData/debug/stepping/lambdaStepInline.kt:1]: Cleaned it up a little. Will just have to eat it and adjust line numbers if needed :laughing:

File: compiler/testData/debug/stepping/lambdaStepInline.kt (hunk 3 above)
Comment (kandersen) [reply compiler/testData/debug/stepping/lambdaStepInline.kt:11]: We can race PR's ;)

**Comment by max-kammerer** [comment]
@yanex Please take a look

File: compiler/testData/debug/stepping/lambdaStepInline.kt [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+// FILE: test.kt
+inline fun foo(stringMaker: () -> String): String {
+    return stringMaker()
+}
+
+fun box(): String {
+    foo { "OK "}
+    foo {
+        "OK"
+        // This line is hit by the old backend after line 13
```
Comment (max-kammerer) [inline compiler/testData/debug/stepping/lambdaStepInline.kt:11]: @kandersen I bit confused by meaning of "after line 13"

Comment (max-kammerer) [reply compiler/testData/debug/stepping/lambdaStepInline.kt]: It was wrong linenumber? Or 13 is correct and i didn't caught the meaning?

Comment (kandersen) [reply compiler/testData/debug/stepping/lambdaStepInline.kt]: Sorry for the premature "closure", I think I might have hit "Resolve" from my phone.

**Comment by max-kammerer** [comment]
Thank you!

File: compiler/testData/debug/stepping/lambdaStepInline.kt (hunk 6 above)
Comment (max-kammerer) [reply compiler/testData/debug/stepping/lambdaStepInline.kt]: Thanks for explanation!

**Pull request merged**